napari/plugins/io.py

```diff
@@ -228,18 +230,40 @@ def save_layers(
     else:
         written = []
 
+    # If written is empty, something went wrong.
+    # Generate a warning to tell the user what it was.
     if not written:
-        # if written is empty, it means no plugin could write the
-        # path/layers combination
-        # we just want to provide some useful feedback
-        warnings.warn(
-            trans._(
-                'No data written! There may be no plugins capable of writing these {length} layers to {path}.',
-                deferred=True,
-                length=len(layers),
-                path=path,
+        if len(layers) == 0:
+            warnings.warn(
+                trans._(
+                    f"No layers were provided for the request to write to {path}",
+                    deferred=True,
+                    path=path,
+                )
+            )
+        elif plugin or _writer:
```

Comment:
this is fine for now, and better than what we currently have, but passing `plugin` or `_writer` isn't the only way that npe2 would be used here.  it's fine to call `save_layers` without either parameter, and it would technically _still_ be nice to know if an npe2 writer was used (but didn't return any files).  So, the error below that says `A plugin could not be found to write these {length} layers to {path}.` could very likely be misleading in certain cases.

let me know if you'd want to tackle that here, or leave it potentially confusing for now

Reply:
If I understand correctly, we could do this by using `npe2.io_utils.write_get_writer` instead of `npe2.io_utils.write` in `_npe2.write_layers`, then bubble up the writer plugin that was used to this function. At first glance that seems fairly straightforward, only needs changes in `napari`, and doesn't change any public API return types, so I think it's probably worth doing.

But if there's some complication I missed, I'd also be fine with what we have here.

Reply:
took this approach.